group/ristretto255.go

```diff
@@ -0,0 +1,179 @@
+package group
+
+import (
+	"crypto"
+	"io"
+
+	h2c "github.com/armfazh/h2c-go-ref"
+
+	r225 "github.com/bwesterb/go-ristretto"
+)
+
+var (
+	Ristretto255 Group = ristrettoGroup{}
+)
+
+type ristrettoGroup struct {
+}
+
+type ristrettoElement struct {
+	p *r225.Point
+}
+
+type ristrettoScalar struct {
+	s *r225.Scalar
+}
+
+func (g ristrettoGroup) NewElement() Element {
+	return &ristrettoElement{
+		p: &r225.Point{},
+	}
+}
+
+func (g ristrettoGroup) NewScalar() Scalar {
+	return &ristrettoScalar{
+		s: &r225.Scalar{},
+	}
+}
+
+func (g ristrettoGroup) Identity() Element {
+	var zero r225.Point
+	zero.SetZero()
+	return &ristrettoElement{
+		p: &zero,
+	}
+}
+
+func (g ristrettoGroup) Generator() Element {
+	var base r225.Point
+	base.SetBase()
+	return &ristrettoElement{
+		p: &base,
+	}
+}
+
+func (g ristrettoGroup) Order() Scalar {
+	panic("not implemented")
+	return nil
+}
+
+func (g ristrettoGroup) RandomElement(r io.Reader) Element {
+	var x r225.Point
+	x.Rand()
+	return &ristrettoElement{
+		p: &x,
+	}
+}
+
+func (g ristrettoGroup) RandomScalar(r io.Reader) Scalar {
+	var x r225.Scalar
+	x.Rand()
+	return &ristrettoScalar{
+		s: &x,
+	}
+}
+
+// Note(caw): this does NOT implement HashToElement as specified in the hash-to-curve draft
+// https://cfrg.github.io/draft-irtf-cfrg-hash-to-curve/draft-irtf-cfrg-hash-to-curve.html#section-appendix.b
+func (g ristrettoGroup) HashToElement(msg, dst []byte) Element {
+	e := g.NewElement()
+
+	expID := h2c.ExpanderDesc{h2c.XMD, uint(crypto.SHA512)}
+	exp, err := expID.Get(dst, 0)
+	if err != nil {
+		panic(err)
+	}
+	data := exp.Expand(msg, 64)
+
+	e.(*ristrettoElement).p.Derive(data)
+	return e
+}
+
+func (g ristrettoGroup) HashToScalar(msg, dst []byte) Scalar {
+	s := g.NewScalar()
+	s.(*ristrettoScalar).s.Derive(msg)
+	return s
+}
+
+func (e *ristrettoElement) IsIdentity() bool {
+	var zero r225.Point
+	zero.SetZero()
+	return e.p.Equals(&zero)
+}
+
+func (e *ristrettoElement) IsEqual(x Element) bool {
+	return e.p.Equals(x.(*ristrettoElement).p)
+}
+
+func (e *ristrettoElement) Add(x Element, y Element) Element {
+	e.p.Add(x.(*ristrettoElement).p, y.(*ristrettoElement).p)
+	return e
+}
+
+func (e *ristrettoElement) Dbl(x Element) Element {
+	return e.Add(x, x)
```

Comment:
I could expose a more efficient double in `go-ristretto`, if desired.

Reply:
Yes, please. :-)

Reply:
https://github.com/bwesterb/go-ristretto/issues/21

Reply:
Done.